source/tools/MatchBin.h

```diff
@@ -0,0 +1,241 @@
+/**
+ *  @note This file is part of Empirical, https://github.com/devosoft/Empirical
+ *  @copyright Copyright (C) Michigan State University, MIT Software license; see doc/LICENSE.md
+ *  @date 2019
+ *
+ *  @file MatchBin.h
+ *  @brief A container that supports flexible tag-based lookup. .
+ *
+ */
+
+
+#ifndef EMP_MATCH_BIN_H
+#define EMP_MATCH_BIN_H
+
+#include <iostream>
+#include <unordered_map>
+#include <unordered_set>
+#include <functional>
+#include <algorithm>
+#include <stdexcept>
+
+#include "../base/assert.h"
+#include "../base/vector.h"
+#include "../tools/IndexMap.h"
+
+namespace emp {
+
+  template<size_t width>
+  struct HammingDistance{
```

Comment:
Looking at how the `std::hash<emp::BitSet<N>>` is implemented, maybe after all it would be a better idea to not have a static `Metric` function and do a non-static `operator()` function. I assume that the people who made that design decision for `std::hash` stuff know best so copying them as closely as possible in this case would probably be a good idea.

Reply:
"I assume that the people who made that design decision for std::hash stuff know best"

Heh.  Definitely not always a good assumption.  They usually have a good reason for what they did do, but haven't always fully considered all of the drawbacks.  There are LOTS of minor examples, and a small number of majorly poor decisions (like specializing `std::vector<bool>`)